### PR TITLE
[TGUI] Temp gun

### PR DESCRIPTION
**Pull request by Contrabang**
## What Does This PR Do
 TGUI's the temperature gun

## Why It's Good For The Game
We are moving away from Topic and into the age of TGUI

## Images of changes

image
image
image
image
image

Emagged and above the previous max temp point (bolded)
image

## Testing

Compiled and ran game, no runtimes. Menu works and responds as expected.

## Changelog
:cl:
tweak: The temperature gun now uses TGUI
/:cl:

### CODE REVIEW

File: code/modules/projectiles/guns/energy/special.dm [hunk 1]
```diff
@@ -590,35 +593,40 @@
 	STOP_PROCESSING(SSobj, src)
 	return ..()
 
-/obj/item/gun/energy/temperature/attack_self(mob/living/user as mob)
-	update_dat()
-	user << browse("<TITLE>Temperature Gun Configuration</TITLE><HR>[dat]", "window=tempgun;size=510x120")
-	onclose(user, "tempgun")
+/obj/item/gun/energy/temperature/attack_self(mob/user)
+	add_fingerprint(usr)
```
Comment (Sirryan2002) [inline code/modules/projectiles/guns/energy/special.dm:597]:
```suggestion
	add_fingerprint(user)
```

File: code/modules/projectiles/guns/energy/special.dm [hunk 2]
```diff
@@ -590,35 +593,40 @@
...
+	data["power_cost"] = powercost
+	return data
+
+/obj/item/gun/energy/temperature/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	if(..())
+		return
+
+	switch(action)
+		if("target_temperature")
+			target_temperature = clamp(text2num(params["target_temperature"]) + T0C, min_temp, max_temp) //Retrieved as a celcius number, convert to kelvin
```
Comment (Sirryan2002) [inline code/modules/projectiles/guns/energy/special.dm:621]:
```suggestion
	if(action == "target_temperature")
			target_temperature = clamp(text2num(params["target_temperature"]) + T0C, min_temp, max_temp) //Retrieved as a celcius number, convert to kelvin
```

File: code/modules/projectiles/guns/energy/special.dm [hunk 3]
```diff
@@ -569,16 +569,19 @@
...
-	var/powercost = ""
-	var/powercostcolor = ""
-	var/dat = ""
+	// Measured in Kelvin
+	var/temperature = T20C
+	var/target_temperature = T20C
+	var/min_temp = 0
+	var/max_temp = 500
+
+	var/powercost = "Low"
```
Comment (Sirryan2002) [inline code/modules/projectiles/guns/energy/special.dm:583]: this method of displaying power cost should be moved out of DM and into the TGUI menu

File: tgui/packages/tgui/interfaces/TempGun.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import { toFixed } from 'common/math';
+import { useBackend } from '../backend';
+import { LabeledList, NumberInput, Section, AnimatedNumber, Box } from '../components';
+import { Window } from '../layouts';
+import { round } from 'common/math';
+
+export const TempGun = (props, context) => {
+  const { act, data } = useBackend(context);
+
```
Comment (Sirryan2002) [inline tgui/packages/tgui/interfaces/TempGun.js:9]:
```suggestion
```

File: tgui/packages/tgui/interfaces/TempGun.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            <AnimatedNumber value={(round(temperature, 2))}/>&deg;C
+        </Box>
+        </LabeledList.Item>
+        <LabeledList.Item
+        label="Power Cost">
+        <Box color={Cost2Color(power_cost)}>
+            {power_cost}
+        </Box>
+        </LabeledList.Item>
+        </LabeledList>
```
Comment (Sirryan2002) [inline tgui/packages/tgui/interfaces/TempGun.js:47]: fix your indentation here

File: tgui/packages/tgui/interfaces/TempGun.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            width="50px"
+            onDrag={(e, value) =>
+              act('target_temperature', {
+                target_temperature: value,
+              })
+            }
+          />&deg;C
+        </LabeledList.Item>
+        <LabeledList.Item
+        label="Current Temperature">
```
Comment (Sirryan2002) [inline tgui/packages/tgui/interfaces/TempGun.js:35]:
```suggestion
        <LabeledList.Item label="Current Temperature">
```

File: tgui/packages/tgui/interfaces/TempGun.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        <Box color={Temp2Color(temperature)}
+            bold={temperature > 500-273.15}>
+            <AnimatedNumber value={(round(temperature, 2))}/>&deg;C
+        </Box>
+        </LabeledList.Item>
+        <LabeledList.Item
+        label="Power Cost">
```
Comment (Sirryan2002) [inline tgui/packages/tgui/interfaces/TempGun.js:42]:
```suggestion
        <LabeledList.Item label="Power Cost">
```

**Review by Sirryan2002 — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/TempGun.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
+import { toFixed } from 'common/math';
+import { useBackend } from '../backend';
+import { LabeledList, NumberInput, Section, AnimatedNumber, Box } from '../components';
+import { Window } from '../layouts';
+import { round } from 'common/math';
+
+export const TempGun = (props, context) => {
+  const { act, data } = useBackend(context);
+  const {target_temperature, temperature, max_temp, min_temp} = data;
```
Comment (Sirryan2002) [inline tgui/packages/tgui/interfaces/TempGun.js:9]:
```suggestion
  const { target_temperature, temperature, max_temp, min_temp } = data;
```

**Review by Sirryan2002 — APPROVED** [review]
code looks good

**Review by SteelSlayer — APPROVED** [review]

**Pull request merged**